lib/ansible/module_utils/facts/__init__.py

```diff
@@ -0,0 +1,212 @@
+#       - one file collectors (say, env.py or cmdline.py) can live in collectors/
+#       - collectors with interdeps (Network or Hardware for ex, may live in collectors/network/)
+#       - or could flatten to collectors/linux_hardware.py, collectors/sunos_hardware.py etc
+# IDEA: gather 'tags' in addition to gather 'ids'
+#       - ids identify _a_ fact collector uniquely
+#       - tags can identify multiple facts collectors
+#       - gather tag for 'hardware' for all 'hardware' tagged collectors
+#       - id could be a namespaced tag (ie, 'id:ansible.module_utils.facts.system.env.EnvFactCollector')
+#       - arch/system/dist could also be namespaced tags
+#           - 'system:Linux'
+#           - 'arch: x86_64'
+# IDEA: once Collector api is used, it wouldn't be that hard to add a collect_iter()
+#       that would return a generator that would yield facts
+#       ... top level Collector could 'emit' facts as found (or changed) which would
+#           make it possibly to watch a fact (or attach a callback to be called when changed)
+#            (more useful for controller side _info than client/remote _facts though given the
+#             controler->remote interface is not really async or non-blocking at all)
+# IDEA: parallel/threaded/multiprocess fact collection
+#       ... the collect_iter() approach above would make that easier, but even for blocking
+#           fact collection, a given Collector could choose to run its sub collectors concurrently.
+#           Might improve latency/total time to collect facts, since fact collection is currently very
+#           serial with lots of things that block and can be slow (more or less every run_command() for
+#           ex). In theory fact collection should be entirely 'read-only' (and with Collector api, with
+#           very few side effects) so might be a reasonable place for some concurency.
+# TODO: possibly rename FooCollector to just FooFacts, esp once the existing Facts() class is barebones/unneeded
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from collections import defaultdict
+
+from ansible.module_utils.facts import timeout
+
+
+# FIXME: make sure get_collector_names returns a useful ordering
+# TODO: may need some form of AnsibleFactNameResolver
+# NOTE: This maps the gather_subset module param to a list of classes that provide them -akl
+def get_collector_names(valid_subsets=None,
+                        minimal_gather_subset=None,
+                        gather_subset=None,
+                        aliases_map=None):
+    '''return a set of FactCollector names based on gather_subset spec.
+
+    gather_subset is a spec describing which facts to gather.
+    valid_subsets is a frozenset of potential matches for gather_subset ('all', 'network') etc
+    minimal_gather_subsets is a frozenset of matches to always use, even for gather_subset='!all'
```

Comment:
Note, we have standardized on sphinx format for docstrings.

See, for instance, ansible/plugins/connection/__init__.py for an example.